scipy/signal/_upfirdn_apply.pyx

```diff
@@ -65,18 +72,145 @@ def _output_len(Py_ssize_t len_h,
     return need
 
 
-def _apply(DTYPE_t [:] x, DTYPE_t [:] h_trans_flip, DTYPE_t [:] out,
-                 Py_ssize_t up, Py_ssize_t down):
-    _apply_impl(x, h_trans_flip, out, up, down)
+def _apply(np.ndarray data, DTYPE_t [::1] h_trans_flip, np.ndarray out,
+           Py_ssize_t up, Py_ssize_t down, Py_ssize_t axis):
+    cdef ArrayInfo data_info, output_info
+    cdef Py_ssize_t len_h = h_trans_flip.size
+    cdef DTYPE_t *data_ptr
+    cdef DTYPE_t *filter_ptr
+    cdef DTYPE_t *out_ptr
+    cdef int retval
+
+    # _apply_axis_inner assumes data and out are C-contiguous
+    data = np.ascontiguousarray(data)
+    out = np.ascontiguousarray(out)
+
+    data_info.ndim = data.ndim
+    data_info.strides = <Py_ssize_t *> data.strides
+    data_info.shape = <Py_ssize_t *> data.shape
+
+    output_info.ndim = out.ndim
+    output_info.strides = <Py_ssize_t *> out.strides
+    output_info.shape = <Py_ssize_t *> out.shape
+
+    data_ptr = <DTYPE_t*> data.data
+    filter_ptr = <DTYPE_t*> &h_trans_flip[0]
+    out_ptr = <DTYPE_t*> out.data
+
+    with nogil:
+        retval = _apply_axis_inner(data_ptr, data_info,
+                                   filter_ptr, len_h,
+                                   out_ptr, output_info,
+                                   up, down, axis)
+    if retval == 1 or retval == 2:
+        raise ValueError("_apply_axis_inner failed with code {}".format(retval))
+    elif retval == 3 or retval == 4:
+        raise MemoryError("_apply_axis_inner failed with code {}".format(retval))
```

Comment:
This is because of a peculiarity in the Python C API. We must do this to allow Cython to generate the appropriate C code.